[[ Bug 22577 ]] Ensure device token is returned in the correct form

livecodepanos opened the pull request:
For getting the device token, we were rendering an NSData into a formatted string,
which would use the debug representation. This worked fine until iOS 12.
On iOS 13, the debug representation of the NSData has changed, and does not
display the whole token. So we need to convert the NSData to a hex string
to get the correct representation.

See https://nshipster.com/apns-device-tokens/ for more details.

Tested this patch with both iOS 12.1 and iOS 13.2 SDKs

runrevmark commented on engine/src/mbliphoneapp.mm:
```diff
@@ -529,8 +529,22 @@ - (void)application:(UIApplication *)p_application didReceiveRemoteNotification:
 - (void)application:(UIApplication*)p_application didRegisterForRemoteNotificationsWithDeviceToken:(NSData*)p_device_token
 {
     NSString *t_to_log = [NSString stringWithFormat:@"%s%@%s", "Application: push notification device token (", p_device_token, ")"];
```
This isn't used anymore I don't think?

runrevmark commented on engine/src/mbliphoneapp.mm:
```diff
@@ -529,8 +529,22 @@ - (void)application:(UIApplication *)p_application didReceiveRemoteNotification:
+    NSString *t_registration_text;
     
+    NSUInteger t_length = p_device_token.length;
+    if (t_length == 0)
```
Does this replicate exactly the old format? (Otherwise it will break people's code in the same way as the revised format from debugDescription does).

livecodepanos replied:
hmm, not really. The old format was an alphanumeric string of length 64, bracketed by `< >`, displayed as 8 segments of 8 digits/chars each, separated by spaces.

The new format is just an alphanumeric string of length 64. Since the new format is already in the form people want to use, do we really want to keep the old format? Maybe we could add a release note saying that the token is already in the "final" form, and no further modifications are needed in the users' scripts?

livecodepanos replied:
Example:
Old format: `<965b251c 6cb1926d e3cb366f dfb16ddd e6b9086a 8a3cac9e 5f857679 376eab7C>`

New format: `965b251c6cb1926de3cb366fdfb16ddde6b9086a8a3cac9e5f857679376eab7C`

runrevmark replied:
Okay - so we need to replicate the exact same format - otherwise it will break people's code (for much the same reasons as the change already did!)

runrevmark replied:
(We don't know what people do with the string they get)

livecodepanos replied:
Ok sure 👍

runrevmark commented on engine/src/mbliphoneapp.mm:
```diff
@@ -525,12 +525,48 @@ - (void)application:(UIApplication *)p_application didReceiveRemoteNotification:
 {
-    NSString *t_to_log = [NSString stringWithFormat:@"%s%@%s", "Application: push notification device token (", p_device_token, ")"];
-    NSString *t_registration_text = [NSString stringWithFormat:@"%@", p_device_token];
+    NSString *t_registration_text;
```
I suggest:
```
NSMutableString *t_registration_text = nil;
NSUInteger t_length = p_device_token.length;
const unsigned char *t_bytes = (const unsigned char *)p_device_token.bytes;
if (t_length != 0)
{
    t_registration_text = [[NSMutableString alloc] init];
    [t_registration_text appendString: @"<"];
    for(int i = 0; i < t_length; i++)
    {
        if (i > 0 && i % 8 == 0)
        {
            [t_registration_text appendString: @" "];
        }
        [t_registration_text appendFormat: @"%02x", t_bytes[i]];
    }
    [t_registration_text appendString: @">"];
}
````

runrevmark replied:
Then after the MCStringCreateWithCFStringRef do:
```
[t_registration_text release];
```

livecodepanos committed:
[[ Bug 22577 ]] Ensure device token is returned in the correct form

For getting the device token, we were rendering an NSData into a formatted string,
which would use the debug representation. This worked fine until iOS 12.
On iOS 13, the debug representation of the NSData has changed, and does not
display the whole token. So we need to convert the NSData to a hex string
to get the correct representation.

runrevmark commented:
@livecode-vulcan review ok 7afe7f7

livecode-vulcan commented:
💙 review by @runrevmark ok 7afe7f7deb8993625ec2f3b4e50ab99da1edf509

livecode-vulcan commented:
😎 test success 7afe7f7deb8993625ec2f3b4e50ab99da1edf509

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/717)
* try-community-armv7-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-ndk16r15/builds/126)
* try-community-arm64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-ndk16r15/builds/126)
* try-community-x86-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-ndk16r15/builds/126)
* try-community-x86_64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-ndk16r15/builds/126)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2243)
* try-community-universal-ios-iphoneos13.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos13.2/builds/18)
* try-community-universal-ios-iphonesimulator13.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator13.2/builds/18)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2272)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2259)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2258)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2292)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2232)

livecodepanos committed:
Merge pull request #7263 from livecodepanos/bugfix-22577

[[ Bug 22577 ]] Ensure device token is returned in the correct form

livecodepanos commented:
Hmm, when this PR was merged, GitHub had some issues according to githubstatus.com

This would explain why this PR is marked as "Closed" (I did not remember closing it) but still says it is merged!
